src/coreclr/vm/i386/asmhelpers.asm

```diff
@@ -1290,14 +1290,17 @@ _GenericCLRToCOMCallStub@0 proc public
 
     ; Get pCLRToCOMCallInfo for return thunk
     mov         ecx, [ebx + CLRToCOMCallMethodDesc__m_pCLRToCOMCallInfo]
+    ; Get size of arguments to pop
+    movzx       ecx, word ptr [ecx + CLRToCOMCallInfo__m_cbStackPop]
 
     STUB_EPILOG_RETURN
 
-    ; Tailcall return thunk
-    jmp [ecx + CLRToCOMCallInfo__m_pRetThunk]
+    ; Pop the incoming arguments off the stack and push back the return
+    ; address on the last slot
+    lea         esp, [esp + ecx + 4]
+    xor         ecx, -1
+    push        dword ptr [esp + ecx - 3]
```

Comment:
Can the stack at negative offset be overwritten between `lea         esp` and `push`, e.g. using sampling profiler? I do not think there is a concept of red zone on x86.

Reply:
Hmm, fair point. I will have to think more about it and perhaps find a different assembly pattern.

Reply:
You should be able to copy the return address to the right spot first and then move esp.

Reply:
Yeah, I just need to reorder it a bit to have the extra temp register but it's definitely doable.

Reply:
Hopefully fixed. It passed the tests on my machine.